services/scheduler/src/main/java/org/mule/service/scheduler/internal/DefaultScheduler.java

```diff
@@ -0,0 +1,294 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.service.scheduler.internal;
+
+import static java.lang.System.nanoTime;
+import static java.util.Collections.synchronizedSet;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static java.util.stream.Collectors.toList;
+
+import org.mule.runtime.core.api.scheduler.Scheduler;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.FutureTask;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+/**
+ * Proxy for a {@link ScheduledExecutorService} that adds tracking of the source of the dispatched tasks.
+ *
+ * @since 4.0
+ */
+public class DefaultScheduler implements Scheduler {
```

Comment:
Does it not make any sense to extend `AbstractExecutorService` and do any wrapping of task, or storing of futures in `newTaskFor` rather than implement all methods here?